applications/IgaApplication/custom_elements/truss_element.cpp

```diff
@@ -0,0 +1,441 @@
+        Values.SetStrainVector(strain_vector);
+
+        mConstitutiveLawVector[point_number]->CalculateValue(Values, TANGENT_MODULUS, rTangentModulusVector[point_number]);
+    }
+}
+
+void TrussElement::CalculateGreenLagrangeStrain(
+    std::vector<double>& rGreenLagrangeVector) const
+{
+    const auto& r_geometry = GetGeometry();
+    auto& r_integration_points = r_geometry.IntegrationPoints();
+    SizeType num_integration_points = r_integration_points.size();
+    if (rGreenLagrangeVector.size() != num_integration_points) {
+        rGreenLagrangeVector.resize(num_integration_points);
+    }
+    for (IndexType point_number = 0; point_number < num_integration_points; ++point_number) {
+        const double l = r_integration_points[point_number].Weight() * norm_2(CalculateActualBaseVector(point_number));
+        const double L = r_integration_points[point_number].Weight() * norm_2(mReferenceBaseVector[point_number]);
+        rGreenLagrangeVector[point_number] = ((l * l - L * L) / (2.00 * L * L));
+    }
+}
+
+/// Returns prestress
+double TrussElement::CalculatePrestressPK2(double reference_a, double actual_a) const
+{
+    // pk2 = cauchy * (L / l);
+    // No integration weight needs to be considered as this would cut off.
+    if (GetProperties().Has(TRUSS_PRESTRESS_PK2)) {
+        return GetProperties()[TRUSS_PRESTRESS_PK2];
+    }
+    else if (GetProperties().Has(TRUSS_PRESTRESS_CAUCHY)) {
+        return GetProperties()[TRUSS_PRESTRESS_CAUCHY] * (reference_a / actual_a);
+    }
+    else {
+        return 0.0;
+    }
+}
+
+/// Returns PK2 stress
+void TrussElement::CalculateStressPK2(
+    std::vector<double>& rStressVector,
+    const ProcessInfo& rCurrentProcessInfo) const
+{
+    const auto& r_geometry = GetGeometry();
+    std::vector<double> green_lagrang_strains(GetGeometry().size());
+    CalculateGreenLagrangeStrain(green_lagrang_strains);
+
+    Vector temp_strain = ZeroVector(1);
+    Vector temp_stress = ZeroVector(1);
+
+    const double num_integration_points = r_geometry.IntegrationPointsNumber();
+    if (rStressVector.size() != num_integration_points) {
+        rStressVector.resize(num_integration_points);
+    }
+
+    for (IndexType point_number = 0; point_number < num_integration_points; ++point_number) {
+        ConstitutiveLaw::Parameters Values(r_geometry, GetProperties(), rCurrentProcessInfo);
+        temp_strain[0] = green_lagrang_strains[point_number];
+        Values.SetStrainVector(temp_strain);
+        Values.SetStressVector(temp_stress);
+        mConstitutiveLawVector[point_number]->CalculateMaterialResponse(Values, ConstitutiveLaw::StressMeasure_PK2);
+
+        const double prestress = CalculatePrestressPK2(
+            norm_2(mReferenceBaseVector[point_number]), norm_2(CalculateActualBaseVector(point_number)));
+        temp_stress[0] += prestress;
+        rStressVector[point_number] = temp_stress[0];
```

Comment:
General question:
As I see it this represents the values used for post-processing to be visualized on each GP, right?
But it already includes the integration weights from `CalculateGreenLagrangeStrain` ...., meaning it represents an integrated entity.
Isn't this contradicting?

In the membrane I write the respective stress tensors to each GP without multiplying them with the integration weight. I might be wrong ....

Reply:
To be honest I am also a bit confused here. In the case of converting PK2 to Cauchy and back we divide l/L or vice versa. In this case it doesn't matter if we include the whole length or just the difference. However in the GL strain we have (l^2-L^2)/(2*L^2) and in this case it would matter if we take the complete length of the segment or just the difference.

I am not sure which one is the right one. I will be happy to discuss this, thanks for pointing this out.

Reply:
as this is just post-processing I suggest testing a simple 1dof tension simulation and check the strain